tsdb/index/postings.go

```diff
@@ -151,6 +151,37 @@ func (p *MemPostings) LabelValues(_ context.Context, name string) []string {
 	return values
 }
 
+// LabelValuesFiltered returns values for which the filter function returns true.
+func (p *MemPostings) LabelValuesFiltered(ctx context.Context, name string, filter func(string) bool) []string {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	var values []string // Not preallocated because we don't know what proportion of values will match.
+	//  We call the filter function in batches from a slice, because that goes 25% faster than calling it one at a time.
+	batchSize := 100000
```

Comment:
Hum, this is interesting, how did you come to this conclusion? Is it because branch prediction works better when not iterating the map? 
OTOH, after processing the first batch we could check the proportion of values filtered out. If we didn't filter out anything (or very little), we can grow the `values` to `len(p.m[name])`.

Reply:
Many hours of comparing before and after, and trying to figure out what the slow part was.
Branch prediction is certainly a possibility, but I haven't gathered any data to support that.

Reply:
Thinking about it, if the cause is branch prediction then this is a false positive from the benchmark.
There is no reason to expect a regular pattern in the label sets of real metrics, across thousands of targets.